starttls/starttls.go

```diff
@@ -34,39 +34,56 @@ type connectResult struct {
 	err   error
 }
 
-func tlsConfigForConnect(connectName, clientCert, clientKey string) (*tls.Config, error) {
+func tlsConfigForConnect(connectName, clientCert, clientKey string) (*tls.Config, **tls.CertificateRequestInfo, error) {
 	conf := &tls.Config{
 		// We verify later manually so we can print results
 		InsecureSkipVerify: true,
 		ServerName:         connectName,
 		MinVersion:         tls.VersionSSL30,
 	}
 
+	var err error
+	var cert tls.Certificate
+
 	if clientCert != "" {
 		keyFile := clientCert
 		if clientKey != "" {
 			keyFile = clientKey
 		}
 
-		cert, err := tls.LoadX509KeyPair(clientCert, keyFile)
+		cert, err = tls.LoadX509KeyPair(clientCert, keyFile)
 		if err != nil {
-			return nil, fmt.Errorf("unable to read client certificate/key: %s\n", err)
+			return nil, nil, fmt.Errorf("unable to read client certificate/key: %s\n", err)
 		}
 
+		// Required even if we set fallback, because of bug in Go 1.8.0 (fixed in 1.8.1)
 		conf.Certificates = []tls.Certificate{cert}
 	}
 
-	return conf, nil
+	cri := setGetClientCertificateCallback(conf, &cert)
+	return conf, cri, nil
+}
+
+func setGetClientCertificateCallback(conf *tls.Config, cert *tls.Certificate) **tls.CertificateRequestInfo {
+	var captured *tls.CertificateRequestInfo
+
+	conf.GetClientCertificate = func(cri *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		captured = cri
+		return cert, nil
+	}
+
+	return &captured
 }
 
 // GetConnectionState connects to a TLS server, returning the connection state.
 // Currently, startTLSType can be one of "mysql", "postgres" or "psql", or the
 // empty string, which does a normal TLS connection. connectTo specifies the
-// address to connect to. connectName sets SNI.  connectAs sets db username, smtp ehlo
-// connectCert and connectKey are client certs
-func GetConnectionState(startTLSType, connectName, connectTo, identity, clientCert, clientKey string, timeout time.Duration) (*tls.ConnectionState, error) {
-	var state *tls.ConnectionState
+// address to connect to. connectName sets SNI. connectIdentity sets DB username,
+// SMTP EHLO. connectCert and connectKey are client cert/key.
+func GetConnectionState(startTLSType, connectName, connectTo, identity, clientCert, clientKey string, timeout time.Duration) (*tls.ConnectionState, *tls.CertificateRequestInfo, error) {
```

Comment:
looks like you changed the comment to say `connectIdentity`, but the variable is still called `identity`